### PR TITLE
Started turning The docker install script, into an idempotent install.

### PR DESCRIPTION
**Pull request by ChrisForti**
- Was not sure how to, or if i even needed to check for `apt-get update`. I feel like an update is already an idempotent command? So i just included with the creation of the encryption certs.

### CODE REVIEW

File: dockers-apt-repo.sh [hunk 1]
```diff
@@ -1,17 +1,40 @@
...
-# Add Docker's official GPG key:
-sudo apt-get update
-sudo apt-get install ca-certificates curl
-sudo install -m 0755 -d /etc/apt/keyrings
-sudo curl -fsSL https://download.docker.com/linux/debian/gpg -o /etc/apt/keyrings/docker.asc
-sudo chmod a+r /etc/apt/keyrings/docker.asc
 
-# Add the repository to Apt sources:
+# Creating encryption certificates
+if (  which ca-certificates curl )
```
Comment (jasondoze) [inline dockers-apt-repo.sh:3]: This is 2 different programs, why not check for each individually?

File: dockers-apt-repo.sh [hunk 2]
```diff
@@ -1,17 +1,40 @@
...
+if (  which gpg )
+then
+  echo " gpg key already exist "
+else 
+  echo " Installing the keyring program "
+  sudo install -m 0755 -d /etc/apt/keyrings 
+  echo " Configuring gpg keyring "
+  sudo curl -fsSL https://download.docker.com/linux/debian/gpg -o /etc/apt/keyrings/docker.asc
+  echo " Modifying gpg file permissions "
+  sudo chmod a+r /etc/apt/keyrings/docker.asc
```
Comment (jasondoze) [inline dockers-apt-repo.sh:22]: Can you explain what this conditional is doing.

File: dockers-apt-repo.sh [hunk 3]
```diff
@@ -1,17 +1,40 @@
...
 
 # Then install docker's latest
-` sudo apt-get install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin`
\ No newline at end of file
+if ( which docker )
+then 
+  echo "Dockers latest already installed"
+else
+  echo "Installing docker"
+  sudo apt-get install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin
```
Comment (jasondoze) [inline dockers-apt-repo.sh:39]: You can just use apt install instead of apt-get, look up the diff between the two.

File: dockers-apt-repo.sh [hunk 4]
```diff
@@ -1,17 +1,40 @@
...
+  sudo curl -fsSL https://download.docker.com/linux/debian/gpg -o /etc/apt/keyrings/docker.asc
+  echo " Modifying gpg file permissions "
+  sudo chmod a+r /etc/apt/keyrings/docker.asc
+fi
+
+# Add the repository to Apt sources
+# Not sure how to put this in a script
 echo \
-  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/debian \
+  "deb [GNU=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/debian \
```
Comment (jasondoze) [inline dockers-apt-repo.sh:28]: I think once you understand what is happening in the previous conditional, this part will make sense.

File: dockers-apt-repo.sh (hunk 3 above)
Comment (jasondoze) [reply dockers-apt-repo.sh:39]: Break this into chunks. Take what you have gotten from this site and decide what are your dependencies, what directories and permissions need to exist for you to get the outcome you are looking for? Then break all those into individual conditionals.

It's ok to have all the docker deps in one conditional in my opinion. Also do any of these installs need a `-y`?

Comment (jasondoze) [reply dockers-apt-repo.sh:39]: Next up, write a script to connect to your pi, then run this script in the pi.

**Review by jasondoze — CHANGES_REQUESTED** [review]

File: dockers-apt-repo.sh (hunk 1 above)
Comment (ChrisForti) [reply dockers-apt-repo.sh:3]: I thought so too, but from the docs It looks like an argument.
https://docs.docker.com/engine/install/debian/

File: dockers-apt-repo.sh (hunk 2 above)
Comment (ChrisForti) [reply dockers-apt-repo.sh:22]: Installing, configuring, and setting permissions for gpg keyring.

File: dockers-apt-repo.sh (hunk 1 above)
Comment (jasondoze) [reply dockers-apt-repo.sh:3]: when in doubt test it in the terminal... which whatever

File: dockers-apt-repo.sh (hunk 2 above)
Comment (jasondoze) [reply dockers-apt-repo.sh:22]: So you check if gpg is installed and if not do all this stuff. What happens if gpg is installed and the keyring fails?

**Pull request closed without merging**